tests/CMakeLists.txt

```diff
@@ -36,6 +39,9 @@ add_executable(
   algo_test/Vstack_test.cpp
 )
 
+# Add the preprocessor definition for CYTNX_TEST_DATA_DIR
+target_compile_definitions(test_main PUBLIC CYTNX_TEST_DATA_DIR="${CYTNX_TEST_DATA_DIR}")
```

Comment:
Is it fine to set the visibility to `PRIVATE`?

Doesn't these lines also need to be added into `tests/gpu/CMakeLists.txt`?

Reply:
I am not an expert on CMAKE (this is the first time I use it), but I gather here PUBLIC means that other targets that link with test_main will also get that symbol defined.  I guess there are no such targets, so PRIVATE is fine.

However, this definition is, in principle, not specific to `test_main`, but anything that gets compiled in this directory.  So would it be even better to use
```
add_compile_definitions(CYTNX_TEST_DATA_DIR="${CYTNX_TEST_DATA_DIR}")
```
?

I missed `tests/gpu/CMakeLists.txt` (I haven't tried compiling the gpu tests yet).  I can add it there too.

Reply:
I tried. `add_compile_definitions()` doesn't propagate to the subdirectories. `add_compile_definitions` only affects the targets defined in the same cmake file, so it is the same with `target_compile_definitions` in this case. I feel that `target_compile_definitions` is safer because it will not cause a surprise result if someone adds a new target in the same camke file in the future.

Reply:
Did you put `add_compile_definitions` at the right place?  According to the cmake documentation, it is a directory property, so it will apply to subdirectories too.  But it needs to be set before the `test_main` executable is added:
```CMakeLists.txt

if(BACKEND_TORCH)
else()

# Define the path to the test data directory
set(CYTNX_TEST_DATA_DIR "${CMAKE_CURRENT_SOURCE_DIR}/test_data_base")

# Add the preprocessor definition for CYTNX_TEST_DATA_DIR
add_compile_definitions(CYTNX_TEST_DATA_DIR="${CYTNX_TEST_DATA_DIR}")

add_executable(
  test_main
  test_tools.cpp
   ...
```

Works for me.

I think this is probably the best solution? If someone does add a new target then it is quite likely that it will want access to the test data directory.  If not, having a defined symbol `CYTNX_TEST_DATA_DIR` is not going to cause a problem.  And anyway, when they add the new target to `CMakeLists.txt`, they could choose to add it either before or after the `add_compile_definitions()` statement, to either get the symbol or not.

Reply:
Ok, the tests compile and run on gpu.  I am getting a bunch of GPU test failures though,
```
[  FAILED  ] 40 tests, listed below:
[  FAILED  ] NetworkTest.gpu_Network_dense_no_order
[  FAILED  ] NetworkTest.gpu_Network_dense_find_optimal
[  FAILED  ] NetworkTest.gpu_Network_dense_order_line
[  FAILED  ] NetworkTest.gpu_Network_dense_specified_order
[  FAILED  ] NetworkTest.gpu_Network_dense_reuse
[  FAILED  ] NconTest.gpu_ncon_default_order
[  FAILED  ] NconTest.gpu_ncon_specified_order
[  FAILED  ] NconTest.gpu_ncon_optimal_order
[  FAILED  ] ContractTest.gpu_Contract_denseUt_optimal_order
[  FAILED  ] ContractTest.gpu_Contract_denseUt_default_order
[  FAILED  ] ContractTest.gpu_Contract_denseUt_specified_order
[  FAILED  ] ContractTest.gpu_Contract_denseUt_optimal_specified_order
[  FAILED  ] ContractTest.gpu_Contracts_denseUt_optimal_order
[  FAILED  ] ContractTest.gpu_Contracts_denseUt_default_order
[  FAILED  ] ContractTest.gpu_Contracts_denseUt_specified_order
[  FAILED  ] ContractTest.gpu_Contracts_denseUt_optimal_specified_order
[  FAILED  ] BlockUniTensorTest.gpu_Trace
[  FAILED  ] DenseUniTensorTest.gpu_Trace
[  FAILED  ] ExpM.gpu_ExpM_test
[  FAILED  ] Lanczos_Gnd.gpu_Bk_Lanczos_Gnd_test
[  FAILED  ] Arnoldi.gpu_which_LM_test
[  FAILED  ] Arnoldi.gpu_which_LR_test
[  FAILED  ] Arnoldi.gpu_which_LI_test
[  FAILED  ] Arnoldi.gpu_which_SM_test
[  FAILED  ] Arnoldi.gpu_which_SR_test
[  FAILED  ] Arnoldi.gpu_which_SI_test
[  FAILED  ] Arnoldi.gpu_mat_type_real_test
[  FAILED  ] Arnoldi.gpu_k1_test
[  FAILED  ] Arnoldi.gpu_k_max
[  FAILED  ] Arnoldi.gpu_smallest_dim
[  FAILED  ] Svd.gpu_dense_one_elem
[  FAILED  ] Svd.gpu_dense_nondiag_test
[  FAILED  ] Svd.gpu_U1_zeros_test
[  FAILED  ] Gesvd.gpu_dense_one_elem
[  FAILED  ] Gesvd.gpu_dense_nondiag_test
[  FAILED  ] linalg_Test.gpu_BkUt_Svd_truncate3
[  FAILED  ] linalg_Test.gpu_BkUt_Qr1
[  FAILED  ] linalg_Test.gpu_BkUt_expM
[  FAILED  ] linalg_Test.gpu_DenseUt_Gesvd_truncate
[  FAILED  ] linalg_Test.gpu_DenseUt_Svd_truncate
```

I looked a bit closer at the first one, error is coming from here: https://github.com/Cytnx-dev/Cytnx/blob/384bafaeee4bd99cf28b3999bf0e5ea72ac4ee26/src/UniTensor_base.cpp#L412
called from: https://github.com/Cytnx-dev/Cytnx/blob/384bafaeee4bd99cf28b3999bf0e5ea72ac4ee26/include/UniTensor.hpp#L3432

I didn't look any further.  Is this expected?

Reply:
> And I think this should automatically get propagated to the gpu tests? I need to set up a cuda build environment to test it.

Aha, you are correct. `add_compile_definitions` should be put before `add_subdirectory`. I didn't notice that.

Reply:
> I didn't look any further. Is this expected?

I think it's not expected.

Reply:
I guess the tests failed because your base branch is `master`. Some bugs are fixed on the `dev-master` branch only.